### PR TITLE
bpo-38307:completes the Stack implementation to yield ending line for each class.

### PR DESCRIPTION
**Pull request by kebab-mai-haddi**
38307: Provide class' end line as well in order to provide the scope of the class in a module.

https://bugs.python.org/issue38307

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@avisrivastava254084

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by avisrivastava254084** [comment]
@the-knights-who-say-ni  I have done the CLA. When will it get updated?

**Comment by brandtbucher** [comment]
Thanks @avisrivastava254084, and welcome to CPython!

Regarding the CLA, the link the bot provided says the following:

> After signing the CLA, please **wait at least one US business day** and then check the status by going to the check-python-cla website. The check will also be run automatically the next time you push changes to your PR.

It's my understanding that these signatures are processed by a human at the PSF, so it probably won't happen on a Saturday night / Sunday morning!

Do you mind updating the PR title to with "bpo-" in front of the issue number, so this can be linked appropriately?

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: Please remove all these whitespace lines that you've added.

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: This `__init__` method isn't needed, since all it does is import `inspect` into the function's scope.

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: Please don't reformat this line. It's fine as-is!

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: Please revert this whitespace change as well. Below also...

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: Please revert this.

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: Unidiomatic, sure. But fine as-is, and unrelated to the change. Please revert.

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: Please remove this rather than just commenting it out. It makes for a better diff.

```suggestion

```

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: This comment belongs in a commit message, not in the source code. The old comment is fine.

```suggestion
    stack = Stack()  # Initialize stack of (class, indent) pairs.
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
@avisrivastava254084: this is a very noisy patch. The real changes only affect about 8 lines, but you've changed over 30. Please revert all of the unrelated formatting/whitespace changes (as well as a few others) that I've highlighted in the review below.

Do you mind explaining why this is necessary? In particular, I'm concerned about:

- ...why we need `endline`. Do you have a specific use case?
- ...if `frames[1].frame.f_locals['start'][0] - 1` is the correct way to calculate this.
- ...if doing it in `__delitem__` of a custom `list` subclass is the right way to do this (I'm almost certain it's not).

At a minimum, this will need tests and a [NEWS entry](https://blurb-it.herokuapp.com/).

**Comment by avisrivastava254084** [comment]
@brandtbucher 
Hi! Sorry for a noisy patch, had to work asap.

> .why we need endline. Do you have a specific use case?
So that the class' scope is known. Right now, class' start line is known but not the end line. If we have the end line, we can know the modules imported in b/w start and end. So, we can then deduce that class A used modules: B and C. In this way, we can say that A is dependent on B and C. 

This is useful in drawing and generating UML diagrams, especially in knowing which class is dependent on which classes.

I"ll make the rest of the changes as suggested.

**Comment by avisrivastava254084** [comment]
>...if frames[1].frame.f_locals['start'][0] - 1 is the correct way to calculate this.
I'd want your suggestions or some guiding material that you think I should go through.
>...if doing it in __delitem__ of a custom list subclass is the right way to do this (I'm almost certain it's not).
Same as above.

**Comment by brandtbucher** [comment]
Rather than subclassing `list` and overriding `__delitem__`, just doing `stack[-1].endline = start - 1` before each `del stack[-1]` should suffice. Also, in keeping consistency with other places in the stdlib where this info is used, I would lean towards the spelling `end_lineno`.

A couple of other things: *if* this is accepted, it will need tests, updated docs, and a [NEWS entry](https://blurb-it.herokuapp.com/). All of the classes and comments in this file should also reference the new attribute. And please change the title of this PR to start with "bpo-38307:" so that the bot (and humans) can find your issue!

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@kebab-mai-haddi

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by kebab-mai-haddi** [comment]
@brandtbucher 
Apologies for the delay. I have made the changes at the code level as you suggested.

I was wondering if you could help me with the "CLA Not Signed" issue?
I changed my username from avisrivastava254084 to kebab-mai-haddi.

**Comment by kebab-mai-haddi** [comment]
@brandtbucher 
I have made the changes. Is the current implementation alright? If so, can you guide me to the testing phase?

File: Lib/pyclbr.py
Comment (brandtbucher) [inline Lib/pyclbr.py]: Why did you switch to using `setattr` here? This should be fine, right?

```suggestion
                    stack[-1][0].end_lineno = start[0] - 1
```

Also on the two other lines, below...

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Again, this still needs tests (in `Lib/test/test_pyclbr.py`) and [a NEWS entry](https://blurb-it.herokuapp.com/). The documentation and docstrings will also need to be updated (basically, anywhere the current `lineno` attribute is referenced).

Regarding the CLA, I think you will need to link your new GitHub username to your BPO account and re-sign the CLA. Not sure, though.

Also, just so you don't get discouraged: it's great that you're taking the time to contribute to CPython, but this module has no current experts or maintainers. I am not able to merge this PR once it's ready, and there is a fair chance that no active core developer would feel comfortable adding a new feature like this - especially without significant demand.

**Comment by csabella** [comment]
@kebab-mai-haddi, on the bug tracker, you should be able to edit your GitHub profile name.  Select 'Your details' on the left and it should take you to a screen where you can update your info.

File: Lib/pyclbr.py
Comment (kebab-mai-haddi) [reply Lib/pyclbr.py]: I have to set the attribute, right? Otherwise, it results into the error: <> has no attribute end_lineno.

Comment (kebab-mai-haddi) [reply Lib/pyclbr.py]: >Also, just so you don't get discouraged: it's great that you're taking the time to contribute to CPython, but this module has no current experts or maintainers. I am not able to merge this PR once it's ready, and there is a fair chance that no active core developer would feel comfortable adding a new feature like this - especially without significant demand.

I received some positive responses on Stackoverflow for this. Don't you think there should be a starting and an end line for this?

Even if you think and abide by to go with the popular demand, should I opt for a survey, etc? Is there some platform where I can ask for votes: "Should there be end_lineno?"

Thank you

**Comment by brandtbucher** [comment]
> I received some positive responses on Stackoverflow for this. Don't you think there should be a starting and an end line for this?
>
> Even if you think and abide by to go with the popular demand, should I opt for a survey, etc? Is there some platform where I can ask for votes: "Should there be end_lineno?"

Sorry, I should have been clearer. I myself am not really opposed to adding this feature; I just don't have the necessary permissions to actually merge this PR, and I've never even seen this module before. Once I deem that it's ready, my next step is usually to request the review of a maintainer or owner of this module who does have the ability to merge it. It doesn't look like we have anybody like that, though.

If you wanted to get somebody's attention, resonding to Steven's questions on the BPO issue would be a great place to start (actually, it should probably happen regardless). If that fizzles out, you could also make a post on [the Python-Ideas mailing list](https://mail.python.org/mailman3/lists/python-ideas.python.org) outlining your use case, and gathering feedback. I would also link this PR and the open BPO issue there, as well.

File: Lib/pyclbr.py
Comment (brandtbucher) [reply Lib/pyclbr.py]: > I have to set the attribute, right? Otherwise, it results into the error: <> has no attribute end_lineno.

They should both be equivalent. You're certain that applying my suggested change results in an error?

Comment (kebab-mai-haddi) [reply Lib/pyclbr.py]: Yes. My bad, apologies. I tried the direct approach to `stack[-1]` instead of `stack[-1][0]`.

**Comment by kebab-mai-haddi** [comment]
@brandtbucher 
I have added the NEWS entry. Regarding tests, I tried the following:
```
t1 = type(o1), o1.name, o1.file, o1.module, o1.lineno, o1.end_lineno
t2 = type(o2), o2.name, o2.file, o2.module, o2.lineno, o2.end_lineno
```
the actual code of which is available in the diff of this branch.

The test is failing as the children being compared are of different types. The error is:
```
======================================================================
ERROR: test_nested (test.test_pyclbr.PyclbrTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/aviralsrivastava/dev/cpython/Lib/test/test_pyclbr.py", line 219, in test_nested
    compare(None, actual, None, expected)
  File "/Users/aviralsrivastava/dev/cpython/Lib/test/test_pyclbr.py", line 212, in compare
    t2 = type(o2), o2.name, o2.file, o2.module, o2.lineno, o2.end_lineno
AttributeError: 'Function' object has no attribute 'end_lineno'

----------------------------------------------------------------------

Ran 6 tests in 5.524s
Merge branch 'endline-in-readmodule-module' of github.com:avisrivastava254084/cpython into endline-in-readmodule-module

FAILED (errors=1)
test test_pyclbr failed
1 test failed again:
    test_pyclbr

== Tests result: FAILURE then FAILURE ==

1 test failed:
    test_pyclbr

1 re-run test:
    test_pyclbr

Total duration: 11 sec 993 ms
Tests result: FAILURE then FAILURE
make: *** [test] Error 2
```

Now, I am failing to understand what and how should I write the test for the new change of `end_lineno`?
Any direction/lead would be greatly appreciated.

**Comment by brandtbucher** [comment]
@kebab-mai-haddi I'm not too familiar with this code, but a cursory scan tells me this:

`pyclbr` creates two types of objects from modules:
- a `Class` for every `class`
- a `Function` for every `def`

I believe your patch only adds the new `end_lineno` attribute to `Class`. Likely, you also want to add it to `Function` as well (for symmetry and completeness).

**Comment by kebab-mai-haddi** [comment]
> @kebab-mai-haddi I'm not too familiar with this code, but a cursory scan tells me this:
> 
> `pyclbr` creates two types of objects from modules:
> 
> * a `Class` for every `class`
> * a `Function` for every `def`
> 
> I believe your patch only adds the new `end_lineno` attribute to `Class`. Likely, you also want to add it to `Function` as well (for symmetry and completeness).

I have done that. If you check [here](https://github.com/python/cpython/pull/16466/files#diff-11e281561648f5044557f8cc970cae6fR211), you will see that if the object is of type `Function`, I am adding `end_lineno`.

Thank you in advance!

**Pull request closed without merging**
